File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(s)<len(p):
            return []
        c1=[0]*26
        ans=[]
        for i in range(len(p)):
            c=ord(p[i])-97
            c1[c]=c1[c]+1
        c2=[0]*26
        k=len(p)
        for i in range(k):
            c=ord(s[i])-97
            c2[c]=c2[c]+1
        if c1==c2:
            ans.append(i-k+1)
        for i in range(k,len(s)):
            c=ord(s[i-k])-97
            c2[c]=c2[c]-1
            d=ord(s[i])-97
            c2[d]=c2[d]+1
            if c2==c1:
                ans.append(i-k+1)
        return ans       

File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.java
class Solution {
    public boolean compare(int []str1,int []str2){
        for(int i=0;i<26;i++){
            if(str1[i]!=str2[i]){
                return false;
            }
        }
        return true;
    }
    public List<Integer> findAnagrams(String s, String p) {
        if(s.length()<p.length()){
            return new ArrayList<>();
        }
        List<Integer> ans=new ArrayList<>();
        int []pIndex=new int[26];
        int []sIndex=new int[26];
        int k=p.length();
        int length=s.length();
        for(int i=0;i<p.length();i++){
            pIndex[p.charAt(i)-'a']=pIndex[p.charAt(i)-'a']+1;
        }
        for(int i=0;i<k;i++){
            sIndex[s.charAt(i)-'a']=sIndex[s.charAt(i)-'a']+1;
        }
        boolean flag=compare(pIndex,sIndex);
        if(flag==true){
            ans.add(0);
        }
        for(int i=k;i<length;i++){
            sIndex[s.charAt(i-k)-'a']=sIndex[s.charAt(i-k)-'a']-1;
            sIndex[s.charAt(i)-'a']=sIndex[s.charAt(i)-'a']+1;
            boolean comparedValue=compare(pIndex,sIndex);
            if(comparedValue){
                ans.add(i-k+1);
            }
        }
        return ans;
    }
}